### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by ionkoto**

### CODE REVIEW

File: com.musala.simple.students.db/pom.xml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+						<phase>package</phase>
+						<goals>
+							<goal>shade</goal>
+						</goals>
+						<configuration>
+							<createDependencyReducedPom>false</createDependencyReducedPom>
+							<transformers>
+								<transformer
+									implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
+									<mainClass>com.musala.simple.students.Main</mainClass>
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/pom.xml]: Isn't this the main class from the other project? :)

File: com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/DatabaseFactory.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+	/**
+	 * Creates a new implementation of the {@link Database}.
+	 * 
+	 * @return a new database of the required type
+	 */
+	public Database getDatabase() {
+		Database database = null;
+
+		switch (this.getDbType()) {
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/DatabaseFactory.java]: Formatting ;)

File: com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/DatabaseFactory.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+		this.setUsername(dbProperties.getProperty(DB_USER));
+		this.setPassword(dbProperties.getProperty(DB_PASSWORD));
+	}
+
+	/**
+	 * Creates a new implementation of the {@link Database}.
+	 * 
+	 * @return a new database of the required type
+	 */
+	public Database getDatabase() {
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/DatabaseFactory.java]: The factory is fine, but we can make it better! :) Let's change this method to be named `public static Database createDatabase(DatabaseTypes dbType)` and implement the so-called Builder pattern (often supported by factories). Perform the following steps:

- Remove the implemented constructor, we won't need it.
- The static method `createDatabase(DatabaseTypes dbType)` (previously `getDatabase()`) should internally do the currently implemented logic with the switch statement and return the newly created `Database` object (MySQL database, Mondo database, etc).
- Then, all setters should be renamed and re-declared too - e.g. `setName()` has to become `withName(String name)` and internally the logic should be - we call `withName(dbProperties.getProperty(DB_NAME))` and underneath it does `this.name = name` and then `return this`.
- Keep in mind that making `createDatabase()` static will require all with-something methods to be static.
- Finally, you will have to have a `public Database build()` method which returns `this` instance.
- The getter methods, on the other hand, can be removed - they don't seem to be used anywhere.

The whole point is that, externally, I can do the following call:

```java
Properties dbProperties = FileHelper.readDbPropertiesFile(DatabaseTypes.MongoDb);
Database database = DatabaseFactory.createDatabase(DatabaseTypes.MongoDb).withName(dbProperties.getProperty(DB_NAME))withHost(dbProperties.getProperty(DB_HOST)).withPort(...).withUsername(...).withPassword().build().
```

You can use this github class as a reference: https://github.com/eclipse/smarthome/blob/4204ce06bb28c28e5f711e720f87ef83beff2e27/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/builder/ThingBuilder.java

File: com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+ * 
+ * @author yoan.petrushinov
+ *
+ */
+public class MyMongoDatabase implements Database {
+	private MongoClient mongo;
+	private MongoDatabase database;
+	private MongoCollection<Document> studentsCollection;
+	private String dbName;
+	private Logger logger = LoggerFactory.getLogger(Main.class);
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java]: Every time you declare a logger, it has to be bound to the current class (this.getClass()). In such way, logs will be printed as being emitted by this class. If you leave this logger with `getLogger(Main.class)`, logs from `MyMongoDatabase` will be listed as logged by the `Main` which is incorrect. Please correct this everywhere.

File: com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+				logger.error(String.format("Problem occured while trying to add %s with id %d to the database.\n",
+						student.getName(), student.getId()));
+				logger.error("Stacktrace:", e);
+			}
+		}
+	}
+
+	@Override
+	public void deleteStudentById(int studentId) {
+
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java]: Why there is no implementation for this method? :)

File: com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+		Document query = new Document("id", studentId);
+		query = this.studentsCollection.find(query).first();
+
+		if (query == null) {
+			throw new StudentNotFoundException(ErrorMessage.STUDENT_NOT_EXISTS);
+		}
+
+		List list = new ArrayList(query.values());
+		int id = (int) list.get(1);
+		String name = (String) list.get(2);
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java]: Everywhere you cast directly (String), replace with `Objects.toString(list.get(2))`. In such case you make use of Java 8's built-in anti-null-pointer String prevention. :) Read more about `Objects.toString()`.

File: com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+	 */
+	@Override
+	public Student[] getAllStudentsArr() {
+		List<Student> studentsList = new ArrayList<>();
+		try (MongoCursor<Document> cur = this.studentsCollection.find().iterator()) {
+			while (cur.hasNext()) {
+
+				Document doc = cur.next();
+
+				List list = new ArrayList(doc.values());
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java]: The logic implemented between lines 200 and 204 is used in both `getStudentById()` and `getAllStudentsArr()` methods - extract it in an external method and call it in both places.

File: com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/helpers/FileHelper.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * This is a helper class providing methods for creating, reading and updating
+ * files.
+ * 
+ * @author yoan.petrushinov
+ * 
+ */
+public class FileHelper {
+	private static Logger logger = LoggerFactory.getLogger(Main.class);
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/helpers/FileHelper.java]: Correct logger.

File: com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/Main.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import com.musala.simple.students.db.helpers.ValidationHelper;
+import com.musala.simple.students.db.internal.ErrorMessage;
+import com.musala.simple.students.db.student.Student;
+import com.musala.simple.students.db.student.StudentDataPrinter;
+import com.musala.simple.students.db.student.StudentGroup;
+import com.musala.simple.students.db.student.StudentWrapper;
+
+public class Main {
+	public static void main(String[] args) {
+		Logger logger = LoggerFactory.getLogger(Main.class);
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/Main.java]: Here the logger declaration should remain as it is! ;)

File: com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/Main.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+		BasicConfigurator.configure();
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/Main.java]: What does log4j has to do with slf4j? Do we need this statement and `import org.apache.log4j.BasicConfigurator;` at all?

**Review by alex-kostadinov — CHANGES_REQUESTED** [review]
That's it for now, please correct the changes. Folder migration has worked fine! :)

File: com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/Database.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
+package com.musala.simple.students.db;
```
Comment (alex-kostadinov) [inline com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/Database.java]: Do the followint package changes:
 - `com.musala.simple.students.db.database` - in this new package the `Database`, `DatabaseFactory` and `DatabaseTypes` classes should be moved.
 - `com.musala.simple.students.db.database.impl` - in this new package the Mongo and other implementation classes should be moved.

**Review by alex-kostadinov — CHANGES_REQUESTED** [review]

File: com.musala.simple.students.db/pom.xml (hunk 1 above)
Comment (ionkoto) [reply com.musala.simple.students.db/pom.xml]: Haha good point 👍

**Pull request merged**